api/v1_queues.inc

```diff
@@ -0,0 +1,101 @@
+<?php
+include_once($relPath.'user_is.inc');
+include_once($relPath.'release_queue.inc');
+include_once("exceptions.inc");
+include_once("misc.inc");
+
+// DP API v1 -- Queues
+
+//---------------------------------------------------------------------------
+// queues/rounds/:roundid -- list queues in round
+function api_v1_queues_rounds($method, $data, $query_params)
+{
+    $round = $data[":roundid"];
+    $show = get_enumerated_param($query_params, "show", "all", ["enabled", "populated", "all"]);
+
+    // Fields we  can query on
+    $valid_fields = [
+        "name",
+        "group",
+    ];
+
+    $valid_return_fields = [
+        "id",
+        "round_id",
+        "group",
+        "ordering",
+        "name",
+        "enabled",
+        "populated",
+        "length",
+        "unheld_length",
+        "comment",
+    ];
+
+    if (user_can_see_queue_settings()) {
+        array_push($valid_return_fields, "project_selector", "release_criterion");
+    }
+
+    // restrict to list of desired fields, if set
+    $return_fields = array_get_as_array($query_params, "field", $valid_return_fields);
+    $return_fields = array_intersect($return_fields, $valid_return_fields);
+
+    $name_selector = array_get($query_params, "name", null);
+    $group_selector = array_get($query_params, "group", null);
+
+    $output = [];
+    foreach (fetch_queues_data($round, $show, false, $name_selector, $group_selector) as $queue_data) {
+        array_remove_invalid_fields($queue_data, $return_fields);
+        $output[] = $queue_data;
+    }
+    return $output;
+}
+
+//---------------------------------------------------------------------------
+// queues/:queueid -- list projects in release queue, and show queue info & stats
+function api_v1_queues($method, $data, $query_params)
+{
+    global $Round_for_round_id_;
+    $queue = $data[":queueid"];
+    $unheld_only = (bool) get_flag_value($query_params, "unheld_only");
+    $show_queue = (bool) get_flag_value($query_params, "show_queue");
+    $show_stats = (bool) get_flag_value($query_params, "show_stats");
+    $show_projects = (bool) get_flag_value($query_params, "show_projects");
+
+    $valid_proj_return_fields = [
+        "project_id",
+        "title",
+        "author",
+        "language",
+        "genre",
+        "difficulty",
+        "modified_date",
+    ];
+
+    // restrict to list of desired fields, if set
+    $return_proj_fields = array_get_as_array($query_params, "proj_field", $valid_proj_return_fields);
+    $return_proj_fields = array_intersect($return_proj_fields, $valid_proj_return_fields);
+
+    $output = [];
+
+    if ($show_queue) {
```

Comment:
When using it, this was unexpected. Having the endpoint not return any data on a valid request was super confusing. Upon reflection I wonder if either
1. this should be 3 separate RESTful endpoints:
   * `v1/queues/:queueid` -- $show_queue=true
   * `v1/queues/:queueid/stats` -- $show_stats=true
   * `v1/queues/:queueid/projects` -- $show_projects=true
2. the endpoint should always return queue details ($show_queue=true) and keep the separate args for $show_stats and $show_projects
3. the endpoint should return everything by default and you can select which field you want from ["queue", "stats", "projects"]

The first option feels the most RESTful, but it adds 3x the round trips for the same data. What parts of this endpoint are useful to you as an end-user, Brian? Do you want/need all of this and you'd rather do 1 query instead of 3?

Reply:
Really, I just need the projects part, so perhaps that should be a separate endpoint.

There's definitely some attraction to having a `section` parameter so you could say `&section[]=info&section[]=stats&section[]=projects`, but I see that this does start to get murkier if we ever want to add PUT verb to the endpoint. It's meaningless to write to either the projects or stats section.

So yeah, 3 endpoints sounds like it's the right way to go.

Reply:
Done.